policy/aggregation_type_compress.go

```diff
@@ -0,0 +1,147 @@
+	// returns error if any invalid aggregation type for Timer is found.
+	DecompressForTimer(compressed CompressedAggregationTypes) (AggregationTypes, error)
+	// DecompressForGauge decompresses aggregation types,
+	// returns error if any invalid aggregation type for Gauge is found.
+	DecompressForGauge(compressed CompressedAggregationTypes) (AggregationTypes, error)
+}
+
+type aggregationCompresser struct {
+	bs *bitset.BitSet
+}
+
+// NewAggregationTypeCompressor returns a new AggregationTypeCompressor.
+func NewAggregationTypeCompressor() AggregationTypeCompressor {
+	// NB(cw): If we start to support more than 64 types, the library will
+	// expand the underlying word list itself.
+	return &aggregationCompresser{bs: bitset.New(TotalAggregationTypes)}
+}
+
+func (c *aggregationCompresser) Compress(aggTypes AggregationTypes) (CompressedAggregationTypes, error) {
+	c.bs.ClearAll()
+	for _, aggType := range aggTypes {
+		if !aggType.IsValid() {
+			return EmptyCompressedAggregationTypes, fmt.Errorf("could not compress invalid AggregationType %v", aggType)
+		}
+		c.bs.Set(uint(aggType))
+	}
+	codes := c.bs.Bytes()
+	var res CompressedAggregationTypes
+	for i := 0; i < CompressedSize && i < len(codes); i++ {
+		res[i] = codes[i]
+	}
+	return res, nil
+}
+
+type aggregationDecompresser struct {
+	buf []uint64
+	res []AggregationType
+}
+
+// NewAggregationTypeDecompressor returns a new AggregationTypeDecompressor.
+func NewAggregationTypeDecompressor() AggregationTypeDecompressor {
+	return &aggregationDecompresser{
+		buf: make([]uint64, CompressedSize),
+		res: make([]AggregationType, 0, TotalAggregationTypes),
+	}
+}
+
+func (c *aggregationDecompresser) Decompress(compressed CompressedAggregationTypes) (AggregationTypes, error) {
+	return c.decompress(compressed, validateAllFn, invalidTypeErrorFn)
+}
+
+func (c *aggregationDecompresser) DecompressForGauge(compressed CompressedAggregationTypes) (AggregationTypes, error) {
+	return c.decompress(compressed, validateGaugeFn, invalidGaugeTypeErrorFn)
+}
+
+func (c *aggregationDecompresser) DecompressForCounter(compressed CompressedAggregationTypes) (AggregationTypes, error) {
+	return c.decompress(compressed, validateCounterFn, invalidCounterTypeErrorFn)
+}
+
+func (c *aggregationDecompresser) DecompressForTimer(compressed CompressedAggregationTypes) (AggregationTypes, error) {
+	return c.decompress(compressed, validateTimerFn, invalidTimerTypeErrorFn)
+}
+
+func (c *aggregationDecompresser) decompress(compressed CompressedAggregationTypes, validate validateFn, err errorFn) (AggregationTypes, error) {
+	for i := range compressed {
+		c.buf[i] = compressed[i]
+	}
+
+	bs := bitset.From(c.buf)
+
+	c.res = c.res[:0]
+	for i, e := bs.NextSet(0); e; i, e = bs.NextSet(i + 1) {
+		aggType := AggregationType(i)
+		if !validate(aggType) {
+			return EmptyAggregationTypes, err(aggType)
+		}
+		c.res = append(c.res, aggType)
+	}
+
+	res := make(AggregationTypes, len(c.res))
```

Comment:
This does another allocation every time the decompress function is called. Perhaps we should have two versions of decompress: `Decompress` that always makes a copy so users who don't care that much about performance can use it, and `DecompressInPlace` that doesn't make copies.